controllers/kubeadmconfig_controller.go

```diff
@@ -259,24 +258,21 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, errors.New("Control plane already exists for the cluster, only KubeadmConfig objects with JoinConfiguration are allowed")
 	}
 
-	// Get certificates to improve security of discovery
-	certificates, err := r.getClusterCertificates(cluster)
-	if err != nil {
+	certificates := internalcluster.NewCertificates()
+	if err := certificates.GetCertificates(ctx, r.Client, cluster); err != nil {
 		log.Error(err, "unable to lookup cluster certificates")
 		return ctrl.Result{}, err
 	}
-	if certificates != nil {
-		hashes, err := certs.CertificateHashes(certificates.ClusterCA.Cert)
-		if err != nil {
-			log.Error(err, "Unable to generate certificate hashes")
-			return ctrl.Result{}, err
-		}
-		if hashes != nil {
-			if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
-				config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
-			}
-			config.Spec.JoinConfiguration.Discovery.BootstrapToken.CACertHashes = hashes
+	hashes, err := certificates.GetCertificateByName(internalcluster.ClusterCAName).Hashes()
+	if err != nil {
+		log.Error(err, "Unable to generate Cluster CA certificate hashes")
+		return ctrl.Result{}, err
+	}
+	if hashes != nil {
```

Comment:
If hashes **is** nil, is that ok or do we need to error?

Reply:
resolved over zoom

Reply:
i should mention the resolution:

All certs must exist at this point. A check was added to make sure that is the case. If some certs are missing it is an error.